core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,213 @@
+        Types.NestedField.optional(
+            7, Column.POSITION_DELETE_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.optional(
+            8, Column.EQUALITY_DELETE_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            9, Column.EQUALITY_DELETE_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.optional(10, Column.TOTAL_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(11, Column.LAST_UPDATED_AT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            12, Column.LAST_UPDATED_SNAPSHOT_ID.name(), Types.LongType.get()));
+  }
+
+  public static CloseableIterable<Record> fromManifest(
+      Table table, ManifestFile manifest, Schema recordSchema) {
+    CloseableIterable<? extends ManifestEntry<? extends ContentFile<?>>> entries =
+        CloseableIterable.transform(
+            ManifestFiles.open(manifest, table.io(), table.specs())
+                .select(scanColumns(manifest.content())) // don't select stats columns
+                .liveEntries(),
+            t ->
+                (ManifestEntry<? extends ContentFile<?>>)
+                    // defensive copy of manifest entry without stats columns
+                    t.copyWithoutStats());
+
+    return CloseableIterable.transform(
+        entries, entry -> fromManifestEntry(entry, table, recordSchema));
+  }
+
+  public static void updateRecord(Record toUpdate, Record fromRecord) {
+    toUpdate.set(
+        Column.SPEC_ID.ordinal(),
+        Math.max(
+            (int) toUpdate.get(Column.SPEC_ID.ordinal()),
+            (int) fromRecord.get(Column.SPEC_ID.ordinal())));
+    incrementLong(toUpdate, fromRecord, Column.DATA_RECORD_COUNT);
+    incrementInt(toUpdate, fromRecord, Column.DATA_FILE_COUNT);
+    incrementLong(toUpdate, fromRecord, Column.DATA_FILE_SIZE_IN_BYTES);
+    checkAndIncrementLong(toUpdate, fromRecord, Column.POSITION_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toUpdate, fromRecord, Column.POSITION_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toUpdate, fromRecord, Column.EQUALITY_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toUpdate, fromRecord, Column.EQUALITY_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toUpdate, fromRecord, Column.TOTAL_RECORD_COUNT);
+    if (toUpdate.get(Column.LAST_UPDATED_AT.ordinal()) != null
+        && fromRecord.get(Column.LAST_UPDATED_AT.ordinal()) != null
+        && ((long) toUpdate.get(Column.LAST_UPDATED_AT.ordinal())
+            < (long) fromRecord.get(Column.LAST_UPDATED_AT.ordinal()))) {
+      toUpdate.set(
+          Column.LAST_UPDATED_AT.ordinal(), fromRecord.get(Column.LAST_UPDATED_AT.ordinal()));
+      toUpdate.set(
+          Column.LAST_UPDATED_SNAPSHOT_ID.ordinal(),
+          fromRecord.get(Column.LAST_UPDATED_SNAPSHOT_ID.ordinal()));
+    }
+  }
+
+  public static Record partitionDataToRecord(
```

Comment:
Because the parquet writer expects an Iceberg `Record` and `IndexedRecord` is not an Iceberg `Record`

```
java.lang.ClassCastException: class org.apache.iceberg.PartitionData cannot be cast to class org.apache.iceberg.data.Record (org.apache.iceberg.PartitionData and org.apache.iceberg.data.Record are in unnamed module of loader 'app')
	at org.apache.iceberg.data.parquet.GenericParquetWriter$RecordWriter.get(GenericParquetWriter.java:41)
	at org.apache.iceberg.parquet.ParquetValueWriters$StructWriter.write(ParquetValueWriters.java:580)
	at org.apache.iceberg.parquet.ParquetValueWriters$StructWriter.write(ParquetValueWriters.java:581)
	at org.apache.iceberg.parquet.ParquetWriter.add(ParquetWriter.java:135)
	at org.apache.iceberg.io.DataWriter.write(DataWriter.java:71)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.apache.iceberg.data.PartitionStatsWriterUtil.writePartitionStatsFile(PartitionStatsWriterUtil.java:82)
```

Reply:
Maybe there're other data types we need to convert. E.g. the generic writer for `Date` expects a `LocalDate`, while the internal class for `Date` is an integer.

Reply:
I will take a look. Is there a util class to handle this already? Because isn't it a common problem?

Reply:
I believe `IdentityPartitionConverters` can be used to do the conversion. And perhaps we can have test cases for date/time partition columns?

Reply:
@lirui-apache: You are right about needing extra conversion for few data types. I have added the test case for all the data type as partition now in `TestPartitionStatsWriterUtil#testAllDatatypePartition` and also improved `IdentityPartitionConverters`

Reply:
Good suggestion. Optimized it.